language/move-prover/src/prelude.bpl

```diff
@@ -149,7 +154,7 @@ function {:inline} ReverseValueArray(a: ValueArray): ValueArray {
     )
 }
 function {:inline} SliceValueArray(a: ValueArray, i: int, j: int): ValueArray { // return the sliced vector of a for the range [i, j)
-    ValueArray((lambda k:int :: if k < j-i then v#ValueArray(a)[i+k] else DefaultValue), j-i)
+    ValueArray((lambda k:int :: if 0 <= k && k < j-i then v#ValueArray(a)[i+k] else DefaultValue), (if j-i < 0 then 0 else j-i))
```

Comment:
Same issue as above. If feel concerned we are allowing index out of range in those functions and giving them some default semantics.

There might be reasons I do not understand, so can be convinced otherwise.
@shazqadeer

Reply:
`SliceValueArray` is different because it is used in the spec expression where no one checks if the index is valid. More generally, how are we treating the following weird cases in the spec expression?:
 * vec[-1], vec[0..len(vec)+1], vec[5..1]
 * 1/0 (div-by-zero)

Reply:
Yes, these are serious problems we need to have a solution for. I don't know what that would be but making those functions total and doing some default behavior on index out of range seems dangerous since one could create wrong specs which pass verification.

I suggest to open an issue stating the problem, leave what you have here, but add a TODO.

Reply:
I'm not sure if this is the reason, but functions that can be used in specs can't be guarded by an abort.  So, they have to return SOMETHING.

In general, I think this is the problem of dealing with partial function in specs.

I suggested to Junkil that he just return the Error value now. 

Note: I think it would be more clear to return an explict Error() instead of DefaultValue, since it's really an error.  When I added Error, I left the definition for DefaultValue in case we wanted to change the value that is contained in the invalid part of arrays to something other than Error().  So, it's not the most obvious way to say "return an error value".

The Error() value is not a complete solution to the problem.  We'll have to think a bit about something better.  I thought it would be good to have in there, as documentation if nothing else.

Reply:
I don't have anything better to suggest right now.  But I agree we should all think about it. Regarding Error(), I see DefaultValue as serving the same purpose. Perhaps we can unify the two concepts by making DefaultValue/Error a separate constructor.  We should land the PR and get back to the problem later.

Reply:
So can we then please just keep it as is, open an issue as suggested above (the original reply from Junkil is a good basis for an issue description), and put a TODO here? (Ideally #linked to the issue number).

Once that's done, happy to approve.

Reply:
I created an github issue for this (https://github.com/libra/libra/issues/3194), which is put in the main description of this PR.